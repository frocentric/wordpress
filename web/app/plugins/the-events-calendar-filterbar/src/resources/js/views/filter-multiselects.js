/* global tribe, tribe_dropdowns */
/* eslint-disable no-var, strict */

/**
 * Makes sure we have all the required levels on the Tribe Object.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar = tribe.filterBar || {};

/**
 * Configures Filter Multiselects Object in the Global Tribe variable.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar.filterMultiselects = {};

/**
 * Initializes in a Strict env the code that manages the filter multiselects.
 *
 * @since 5.0.0
 *
 * @param  {PlainObject} $   jQuery
 * @param  {PlainObject} _   Underscore.js
 * @param  {PlainObject} obj tribe.filterBar.filterMultiselects
 *
 * @return {void}
 */
( function( $, _, obj ) {
	'use strict';
	var $document = $( document );

	/**
	 * Selectors used for configuration and setup.
	 *
	 * @since 5.0.0
	 *
	 * @type {PlainObject}
	 */
	obj.selectors = {
		multiselectInput: '[data-js="tribe-filter-bar-c-multiselect-input"]',
	};

	/**
	 * Handle multiselect change event.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object of change event.
	 *
	 * @return {void}
	 */
	obj.handleMultiselectChange = function( event ) {
		var key = event.data.target.attr( 'name' );

		// Return early if name attribute is not set.
		if ( ! key ) {
			return;
		}

		// Return early if an ajax request is already happening.
		if ( tribe.events.views.manager.currentAjaxRequest ) {
			return;
		}

		var location = tribe.filterBar.filters.removeKeyValueFromQuery( window.location, key, true );
		var value = event.data.target.attr( 'value' );

		if ( value ) {
			var uniqueValues = _.uniq( value.split( ',' ) );
			var filteredValues = _.filter( uniqueValues, _.identity );
			filteredValues.forEach( function( filteredValue ) {
				location = tribe.filterBar.filters.addKeyValueToQuery( location, key, filteredValue );
			} );
		}

		tribe.filterBar.filters.submitRequest( event.data.container, location.href );
	};

	/**
	 * Debounce the multiselect change handler.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object of change event.
	 *
	 * @return {void}
	 */
	obj.debouncedHandleMultiselectChange = _.debounce( obj.handleMultiselectChange, 50 );

	/**
	 * Handler for template selection.
	 *
	 * @since  5.0.0
	 *
	 * @param  {PlainObject} state State of selected item.
	 *
	 * @return {jQuery}
	 */
	obj.handleTemplateSelection = function( state ) {
		var $newEl = $( '<span class="select2-selection__choice__text"></span>' );
		$newEl.text( state.text );
		return $newEl;
	};

	/**
	 * Initializes filter multiselect.
	 *
	 * @param  {jQuery}  $multiselectInput jQuery object of multiselect input.
	 * @param  {jQuery}  $container        jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.initMultiselect = function( $multiselectInput, $container ) {
		tribe_dropdowns.dropdown( $multiselectInput, {
			templateSelection: obj.handleTemplateSelection,
		} );
		$multiselectInput
			.on( 'change', { target: $multiselectInput, container: $container }, obj.debouncedHandleMultiselectChange )
			.data( 'select2' ).$container.addClass( 'select2-container--open' );
		$multiselectInput.data( 'select2' ).trigger( 'query', {} );
	};

	/**
	 * Deinitializes filter multiselects.
	 *
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.deinitMultiselects = function( $container ) {
		$container
			.find( obj.selectors.multiselectInput )
			.each( function( index, multiselectInput ) {
				$( multiselectInput )
					.off()
					.select2( 'destroy' );
			} );
	};

	/**
	 * Initializes filter multiselects.
	 *
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.initMultiselects = function( $container ) {
		$container
			.find( obj.selectors.multiselectInput )
			.each( function( index, multiselectInput ) {
				obj.initMultiselect( $( multiselectInput ), $container );
			} );
	};

	/**
	 * Handler for `keydown` event. Prevent `keydown` event from bubbling up.
	 *
	 * @todo: @paulmskim this will be removed once a more permanent solution can be put in place.
	 *
	 * @since  5.0.2
	 *
	 * @param  {Event} event event object for 'keydown' event.
	 *
	 * @return {void}
	 */
	obj.handleKeyDown = function( event ) {
		event.stopPropagation();
	};

	/**
	 * Handler for `focusin` event.
	 *
	 * @todo: @paulmskim this will be removed once a more permanent solution can be put in place.
	 *
	 * @since  5.0.2
	 *
	 * @param  {Event} event event object for 'focusin' event.
	 *
	 * @return {void}
	 */
	obj.handleFocusIn = function( event ) {
		$( event.target ).on( 'keydown', obj.handleKeyDown );
	};

	/**
	 * Handler for `focusout` event.
	 *
	 * @todo: @paulmskim this will be removed once a more permanent solution can be put in place.
	 *
	 * @since  5.0.2
	 *
	 * @param  {Event} event event object for 'focusout' event.
	 *
	 * @return {void}
	 */
	obj.handleFocusOut = function( event ) {
		$( event.target ).off( 'keydown', obj.handleKeyDown );
	};

	/**
	 * Deinitialize filter multiselects JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object for 'beforeAjaxSuccess.tribeEvents' event.
	 *
	 * @return {void}
	 */
	obj.deinit = function( event ) {
		var $container = event.data.container;
		obj.deinitMultiselects( $container );
		$container.off( 'beforeAjaxSuccess.tribeEvents', obj.deinit );
	};

	/**
	 * Initialize filter multiselects JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event}   event      event object for 'afterSetup.tribeEvents' event.
	 * @param  {integer} index      jQuery.each index param from 'afterSetup.tribeEvents' event.
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.init = function( event, index, $container ) {
		obj.initMultiselects( $container );
		$container.on( 'beforeAjaxSuccess.tribeEvents', { container: $container }, obj.deinit );

		/**
		 * @todo: @paulmskim this will be removed once a more permanent solution can be put in place.
		 */
		$container
			.find( '.tribe-events-c-search__input' )
			.on( 'focusin', { container: $container }, obj.handleFocusIn )
			.on( 'focusout', { container: $container }, obj.handleFocusOut );
	};

	/**
	 * Handles the initialization of filter multiselects when Document is ready.
	 *
	 * @since 5.0.0
	 *
	 * @return {void}
	 */
	obj.ready = function() {
		$document.on( 'afterSetup.tribeEvents', tribe.events.views.manager.selectors.container, obj.init );
	};

	// Configure on document ready.
	$( obj.ready );
} )( jQuery, _, tribe.filterBar.filterMultiselects );
