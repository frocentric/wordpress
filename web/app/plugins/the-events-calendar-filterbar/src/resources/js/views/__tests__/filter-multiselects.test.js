describe( 'Filter Multiselects', () => {
	beforeAll( () => {
		String.prototype.className = function() {
			if (
				(
					'string' !== typeof this &&
					! this instanceof String /* eslint-disable-line no-unsafe-negation */
				) ||
				'function' !== typeof this.replace
			) {
				return this;
			}

			return this.replace( '.', '' );
		};

		global.tribe = {};
		require( '../filter-multiselects' );
		global.tribe.filterBar.filters = {};
		global.tribe_dropdowns = {};
	} );

	afterAll( () => {
		delete String.prototype.className;
		delete global.tribe;
		delete global.tribe_dropdowns;
	} );

	describe( 'Selectors', () => {
		test( 'Should match snapshot', () => {
			const selectors = JSON.stringify( tribe.filterBar.filterMultiselects.selectors );
			expect( selectors ).toMatchSnapshot();
		} );
	} );

	describe( 'Handle multiselect change', () => {
		beforeEach( () => {
			global.tribe.events = {
				views: {
					manager: {
						currentAjaxRequest: null,
					},
				},
			};
			global.tribe.filterBar.filters = {
				addKeyValueToQuery: jest.fn().mockImplementation( () => ( {} ) ),
				removeKeyValueFromQuery: jest.fn().mockImplementation( () => ( {} ) ),
				submitRequest: jest.fn(),
			};
		} );

		test( 'Should return early if name attribute is empty', () => {
			// Setup test.
			const event = {
				data: {
					target: $( '<div value="42"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterMultiselects.handleMultiselectChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 0 );
		} );

		test( 'Should return early if an ajax request is already happening', () => {
			// Setup test.
			global.tribe.events = {
				views: {
					manager: {
						currentAjaxRequest: {},
					},
				},
			};
			const event = {
				data: {
					target: $( '<div name="key" value="42"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterMultiselects.handleMultiselectChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 0 );

			// Cleanup test.
			delete global.tribe.events.views.manager;
		} );

		test( 'Should remove key value from query', () => {
			// Setup test.
			const event = {
				data: {
					target: $( '<div name="key"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterMultiselects.handleMultiselectChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 1 );
		} );

		test( 'Should replace key value in query with selected option', () => {
			// Setup test.
			const event = {
				data: {
					target: $( '<div name="key" value="42"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterMultiselects.handleMultiselectChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 1 );
		} );

		test( 'Should replace key value in query with selected options', () => {
			// Setup test.
			const event = {
				data: {
					target: $( '<div name="key" value="42,24"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterMultiselects.handleMultiselectChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 2 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Handle template selection', () => {
		test( 'Should return selected option wrapped in span element', () => {
			// Setup test.
			const state = {
				text: 'Hello world',
			};

			// Test.
			const $template = tribe.filterBar.filterMultiselects.handleTemplateSelection( state );

			// Confirm final states.
			expect( $template ).toMatchSnapshot();
		} );
	} );

	describe( 'Initialize multiselect', () => {
		test( 'Should initialize multiselect', () => {
			// Setup test.
			global.tribe_dropdowns = {
				dropdown: jest.fn(),
			};
			const $multiselectInput = $( '<input />' );
			const $container = $( '<div></div>' );
			$multiselectInput.on = jest.fn().mockImplementation( () => $multiselectInput );
			const addClass = jest.fn();
			const trigger = jest.fn();
			$multiselectInput.data = () => ( {
				$container: {
					addClass: addClass,
				},
				trigger: trigger,
			} );

			// Test.
			tribe.filterBar.filterMultiselects.initMultiselect( $multiselectInput, $container );

			// Confirm final states.
			expect( global.tribe_dropdowns.dropdown.mock.calls.length ).toBe( 1 );
			expect( $multiselectInput.on.mock.calls.length ).toBe( 1 );
			expect( addClass.mock.calls.length ).toBe( 1 );
			expect( trigger.mock.calls.length ).toBe( 1 );
			expect( $multiselectInput.on.mock.calls[ 0 ][ 0 ] ).toBe( 'change' );
		} );
	} );
} );
