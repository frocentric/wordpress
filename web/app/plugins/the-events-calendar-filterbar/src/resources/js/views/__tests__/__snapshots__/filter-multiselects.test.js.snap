// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Multiselects Handle template selection Should return selected option wrapped in span element 1`] = `
jQuery {
  "0": <span
    class="select2-selection__choice__text"
  >
    Hello world
  </span>,
  "length": 1,
}
`;

exports[`Filter Multiselects Selectors Should match snapshot 1`] = `"{\\"multiselectInput\\":\\"[data-js=\\\\\\"tribe-filter-bar-c-multiselect-input\\\\\\"]\\"}"`;
